### ISSUE TITLE
Check if type is jsoncons compatible

### ISSUE THREAD

**Issue by patternoia** [danielaparker/jsoncons#195]
Hi,
I would like to check at run-time if the type I'm working with is json-convertible.
For this I would need to use:
```
if (jsoncons::is_json_type_traits_declared<TestType>::value) {
...
} else {
...
}
```

However it does not work (returns false) for the types I declared the traits with the traits macros (in my case `JSONCONS_PROPERTY_TRAITS_DECL`)

However, if I explicitly specialize `is_json_type_traits_declared`:

```
namespace jsoncons {
template <>
struct is_json_type_traits_declared< TestType> : public std::true_type {};
}
```

Then such a check would return true.

Am I trying to misuse the jsoncons internal functionality? If so, then is there another way to check for compatibility? If not, can this specialization be included into the trait macros?

Thanks

**Comment by danielaparker** [comment]
It's a good question.

A number of the specializations of `json_type_traits` provided in the jsoncons library are based on properties of types, for example, properties of sequence and associative map container types, and apply to all types that have those properties.  We received a couple of issues from users that had user defined classes that also had those properties, but wanted to use their own `json_type_traits` specializations, not the jsoncons library ones. We therefore introduced  the trait class `is_json_type_traits_declared` as a way for allowing users to inform the jsoncons library that the type is already specialized.

The reason we needed `is_json_type_traits_declared`  in the first place is because the `json_type_traits` specializations are _partial_ template class specializations, partial specializations on type `T` in

```c++
template <class Json, class T, class Enable = void>
struct json_type_traits
{
    static constexpr bool is_compatible = false;

    bool is_json(const Json& j)
    {
        return false;
    }

    // ...
};
```
where the template parameter `Json` can be substituted with any type (including a user defined one) that satisfies the requirements for a json value. If it was a full specialization, the jsoncons library wouldn't need a user provided `is_json_type_traits_declared<T>` to know whether type `T` was already specialized; it could figure it out for itself. And we could have freed up  `is_json_type_traits_declared<T>`  to give you what you want.

In retrospect, it may have been better had we defined `json_type_traits` as 

```c++
template <class T, class Enable = void>
struct json_type_traits
{
    static constexpr bool is_compatible = false;

    template <class Json>
    bool is_json(const Json& j)
    {
        return false;
    }

    // ...
};
```
Then we could have defined
```c++
template <class T, class Enable = void>
struct is_json_type_traits_declared : std::true_type
{};

template <class T>
struct is_json_type_traits_declared<T, typename std::enable_if<!json_type_traits<T>::is_compatible>::type> : std::false_type
{};
```
which I think would have served the library's purposes, as well as yours. 

It's too late to make changes to `json_type_traits` or `is_json_type_traits_declared`, but I've been thinking about introducing a new traits class `json_ser_traits`, with some enhancements, and with backwards compatibility with `json_type_traits` . The convenience macros could be transparently changed to generate `json_ser_traits`, and the jsoncons library implementations would support them. So you might get ``is_json_ser_traits_declared`.

Daniel

**Comment by danielaparker** [comment]
`is_json_type_traits_declared<T>` is now specialized for all types that are specialized through the convenience macros on master,

**Comment by danielaparker** [comment]
Included in v0.143.0
